File: client/lib/engine/tribute.dart
import 'types.dart';
import 'card.dart';
import 'team_assigner.dart';

class TributePair {
  final String fromPlayerId;
  final String toPlayerId;
  Card? cardFromLoser;
  Card? cardFromWinner;
  TributePair({required this.fromPlayerId, required this.toPlayerId});
}

List<TributePair> determineTributePairs(
  GameResult gameResult,
  Map<String, String> teams,
  Map<String, List<Card>> hands,
  int maxPlayers,
) {
  if (gameResult.winner == 'draw') return [];

  final loserTeam = gameResult.winner == 'red' ? 'black' : 'red';
  final winners = getTeamMembers(teams, gameResult.winner);
  final losers = getTeamMembers(teams, loserTeam);
  final pairs = <TributePair>[];

  if (maxPlayers == 3) {
    if (gameResult.winner == 'red') {
      for (final loserId in losers) {
        pairs.add(TributePair(fromPlayerId: loserId, toPlayerId: winners.first));
      }
    } else {
      pairs.add(TributePair(fromPlayerId: losers.first, toPlayerId: gameResult.firstFinisherId));
    }
  } else if (maxPlayers == 4) {
    final shuffledWinners = List<String>.from(winners)..shuffle();
    final shuffledLosers = List<String>.from(losers)..shuffle();
    for (var i = 0; i < shuffledWinners.length && i < shuffledLosers.length; i++) {
      pairs.add(TributePair(fromPlayerId: shuffledLosers[i], toPlayerId: shuffledWinners[i]));
    }
  } else if (maxPlayers == 5) {
    final redThreeHolder = _findHolder(hands, teams, Suit.H);
    final diamondHolder = _findHolder(hands, teams, Suit.D);
    final shuffledLosers = List<String>.from(losers)..shuffle();
    final shuffledWinners = List<String>.from(winners)..shuffle();

    if (gameResult.winner == 'red') {
      if (redThreeHolder != null) {
        for (var i = 0; i < 2 && i < shuffledLosers.length; i++) {
          pairs.add(TributePair(fromPlayerId: shuffledLosers[i], toPlayerId: redThreeHolder));
        }
      }
      if (diamondHolder != null && shuffledLosers.length > 2) {
        pairs.add(TributePair(fromPlayerId: shuffledLosers[2], toPlayerId: diamondHolder));
      }
    } else {
      if (redThreeHolder != null) {
        for (var i = 0; i < 2 && i < shuffledWinners.length; i++) {
          pairs.add(TributePair(fromPlayerId: redThreeHolder, toPlayerId: shuffledWinners[i]));
        }
      }
      if (diamondHolder != null && shuffledWinners.length > 2) {
        pairs.add(TributePair(fromPlayerId: diamondHolder, toPlayerId: shuffledWinners[2]));
      }
    }
  }

  return pairs;
}

String? _findHolder(Map<String, List<Card>> hands, Map<String, String> teams, Suit suit) {
  for (final entry in teams.entries) {
    if (entry.value != 'red') continue;
    final hand = hands[entry.key];
    if (hand == null) continue;
    if (hand.any((c) => c.suit == suit && c.rank == Rank.THREE)) {
      return entry.key;
    }
  }
  return null;
}

Card? applyTributeGive(Map<String, List<Card>> hands, TributePair pair) {
  final loserHand = hands[pair.fromPlayerId];
  if (loserHand == null || loserHand.isEmpty) return null;

  final highest = findHighestCard(loserHand);
  if (highest == null) return null;

  loserHand.removeWhere((c) => c.id == highest.id);
  pair.cardFromLoser = highest;
  return highest;
}

bool applyTributeReceive(Map<String, List<Card>> hands, TributePair pair, Card returnCard) {
  final winnerHand = hands[pair.toPlayerId];
  if (winnerHand == null) return false;

  winnerHand.removeWhere((c) => c.id == returnCard.id);
  pair.cardFromWinner = returnCard;

  if (pair.cardFromLoser != null) {
    winnerHand.add(pair.cardFromLoser!);
  }

  final loserHand = hands[pair.fromPlayerId];
  if (loserHand != null) {
    loserHand.add(returnCard);
  }
  return true;
}

File: server/src/engine/tribute.ts
import { Card, GameResult, Team, Suit, Rank, TributePair } from './types';
import { findHighestCard, getSingleCardPower } from './card';
import { getTeamMembers } from './team-assigner';

export function determineTributePairs(
  gameResult: GameResult,
  teams: Map<string, Team>,
  hands: Map<string, Card[]>,
  maxPlayers: number
): TributePair[] {
  if (gameResult.winner === 'draw') return [];

  const loserTeam: Team = gameResult.winner === 'red' ? 'black' : 'red';
  const winners = getTeamMembers(teams, gameResult.winner);
  const losers = getTeamMembers(teams, loserTeam);

  const pairs: TributePair[] = [];

  if (maxPlayers === 3) {
    if (gameResult.winner === 'red') {
      // Red (1 player) wins: 2 black losers each give 1 to red
      for (const loserId of losers) {
        pairs.push({ fromPlayerId: loserId, toPlayerId: winners[0] });
      }
    } else {
      // Black wins: 1 red loser gives to first-finishing black player
      pairs.push({ fromPlayerId: losers[0], toPlayerId: gameResult.firstFinisherId });
    }
  } else if (maxPlayers === 4) {
    // 2 red vs 2 black: 1-to-1 pairing
    const shuffledWinners = [...winners].sort(() => Math.random() - 0.5);
    const shuffledLosers = [...losers].sort(() => Math.random() - 0.5);
    for (let i = 0; i < Math.min(winners.length, losers.length); i++) {
      pairs.push({ fromPlayerId: shuffledLosers[i], toPlayerId: shuffledWinners[i] });
    }
  } else if (maxPlayers === 5) {
    // 2 red vs 3 black
    const redThreeHolder = findRedThreeHolderInTeams(hands, teams, Suit.HEART);
    const diamondThreeHolder = findRedThreeHolderInTeams(hands, teams, Suit.DIAMOND);

    if (gameResult.winner === 'red') {
      // Red wins: 红桃3 receives 2, 方片3 receives 1
      const blackLosers = [...losers].sort(() => Math.random() - 0.5);
      if (redThreeHolder) {
        for (let i = 0; i < 2 && i < blackLosers.length; i++) {
          pairs.push({ fromPlayerId: blackLosers[i], toPlayerId: redThreeHolder });
        }
      }
      if (diamondThreeHolder && blackLosers.length > 2) {
        pairs.push({ fromPlayerId: blackLosers[2], toPlayerId: diamondThreeHolder });
      }
    } else {
      // Black wins: 红桃3 gives 2, 方片3 gives 1
      const blackWinners = [...winners].sort(() => Math.random() - 0.5);
      if (redThreeHolder) {
        for (let i = 0; i < 2 && i < blackWinners.length; i++) {
          pairs.push({ fromPlayerId: redThreeHolder, toPlayerId: blackWinners[i] });
        }
      }
      if (diamondThreeHolder && blackWinners.length > 2) {
        pairs.push({ fromPlayerId: diamondThreeHolder, toPlayerId: blackWinners[2] });
      }
    }
  }

  return pairs;
}

function findRedThreeHolderInTeams(
  hands: Map<string, Card[]>,
  teams: Map<string, Team>,
  suit: Suit
): string | null {
  for (const [playerId, team] of teams) {
    if (team !== 'red') continue;
    const hand = hands.get(playerId);
    if (!hand) continue;
    const has = hand.some((c) => c.suit === suit && c.rank === Rank.THREE);
    if (has) return playerId;
  }
  return null;
}

export function applyTributeGive(
  hands: Map<string, Card[]>,
  pair: TributePair
): Card | null {
  const loserHand = hands.get(pair.fromPlayerId);
  if (!loserHand || loserHand.length === 0) return null;

  const highestCard = findHighestCard(loserHand);
  if (!highestCard) return null;

  const index = loserHand.findIndex((c) => c.id === highestCard.id);
  if (index === -1) return null;

  loserHand.splice(index, 1);
  pair.cardFromLoser = highestCard;
  return highestCard;
}

export function applyTributeReceive(
  hands: Map<string, Card[]>,
  pair: TributePair,
  returnCard: Card
): boolean {
  const winnerHand = hands.get(pair.toPlayerId);
  if (!winnerHand) return false;

  const index = winnerHand.findIndex((c) => c.id === returnCard.id);
  if (index === -1) return false;

  winnerHand.splice(index, 1);
  pair.cardFromWinner = returnCard;

  // Give loser's tribute card to winner
  if (pair.cardFromLoser) {
    winnerHand.push(pair.cardFromLoser);
  }

  // Give winner's return card to loser
  const loserHand = hands.get(pair.fromPlayerId);
  if (loserHand) {
    loserHand.push(returnCard);
  }

  return true;
}

export function executeFullTribute(
  hands: Map<string, Card[]>,
  pairs: TributePair[],
  returnCardChoices: Map<string, Card> // pair index (as string) -> return card
): void {
  for (let i = 0; i < pairs.length; i++) {
    const pair = pairs[i];
    const returnCard = returnCardChoices.get(String(i));
    if (!returnCard) continue;

    applyTributeGive(hands, pair);
    applyTributeReceive(hands, pair, returnCard);
  }
}
